Add progress bar and score percentage as discussed in #102

radotreyes opened the pull request:
Small issue:
<img width="525" alt="screen shot 2018-06-30 at 4 52 25 pm" src="https://user-images.githubusercontent.com/25578179/42130013-c530c774-7c8a-11e8-8860-09f4ca3bafd9.png">
This is what the progress bar looks like right now. There's not anything behaviorally wrong with it, but I just noticed that *when zoomed out*, the red segment appears to be slightly lower than the green segment (it's not, if you check the page source). I'm not sure if I should bother trying to fix this as it's mostly a design issue right now, but if anyone has any suggestions I'm all ears!

radotreyes commented on src/components/Question/Question.js:
```diff
@@ -53,9 +53,34 @@ class Question extends Component {
+    // `score`'s properties ( newScore = { books: ..., } etc. )
+    // vs. initializing it as it is below (line 80)
+    // the result is that the newScore object doesn't actually
+    // get updated
```
Also, when implementing the updated score object, I ran into this weird bug(?). My goal was to remove `newScore` altogether and just contain the entire "newScore" object as an argument to `updateScore()` (line 92) without an additional variable declaration. However, this (lines 75 - 79) happens. 

The app behaves as expected with the changes as they are now, but would just like to know why this happens and if I can indeed minify the code a little as I described.

radotreyes commented on src/components/Question/Question.js:
```diff
@@ -53,9 +53,34 @@ class Question extends Component {
+    // NOTE: possible bug; if we declare `newScore` by initializing to
+    // `score`'s properties ( newScore = { books: ..., } etc. )
+    // vs. initializing it as it is below (line 80)
+    // the result is that the newScore object doesn't actually
```
Also, when implementing the updated score object, I ran into this weird bug(?). My goal was to remove `newScore` altogether and just contain the entire "newScore" object as an argument to `updateScore()` (line 92) without an additional variable declaration. However, this (lines 75 - 79) happens. 

The app behaves as expected with the changes as they are now, but would just like to know why this happens and if I can indeed minify the code a little as I described.

radotreyes commented on src/components/Question/Question.js:
```diff
@@ -53,9 +53,34 @@ class Question extends Component {
+    // return a new score object
+    // NOTE: possible bug; if we declare `newScore` by initializing to
+    // `score`'s properties ( newScore = { books: ..., } etc. )
+    // vs. initializing it as it is below (line 80)
```
Also, when implementing the updated score object, I ran into this weird bug(?). My goal was to remove `newScore` altogether and just contain the entire "newScore" object as an argument to `updateScore()` (line 92) without an additional variable declaration. However, this (lines 75 - 79) happens. 

The app behaves as expected with the changes as they are now, but would just like to know why this happens and if I can indeed minify the code a little as I described.

austintackaberry replied:
Here is a way to update the score immutably:

```javascript
const newScore = {...score, books: score.books.map((book, index) => {
  if (index === bookId) {
    return {
      ...book,
      correct: book.correct + scoreDiff.correct,
      incorrect: book.incorrect + scoreDiff.incorrect,
      chapters: book.chapters.map((chapter, index) => {
        if (index === chapterId) {
          return {
            ...chapter,
            questions: chapter.questions.map((question, index) => {
              if (questionIndex === index) {
                return {
                  answered: true || question.answered,
                  correct: isCorrect,
                };
              }
              return question
            })
          }
        }
        return chapter
      })
    }
  }
  return book
})}
```

In order to use the spread operator we need to add:
`plugins: ['transform-object-rest-spread'],` to the babelrc

austintackaberry replied:
Just thinking out loud...I feel like the best way to structure `score` would be to not store `book.correct` or `book.incorrect`. Also, for each question, instead of `{answered, correct}`, we just have `{status: "correct" }`, `{status: "incorrect" }`, or `{status: "unanswered" }`

Then we should just use functions to calculate everything we need to get out of the data structure. This doesn't have to be a part of this PR...just a thought I had.

radotreyes commented:
Thanks for the feedback! I went ahead and implemented your suggestions and everything still seems fine. But suddenly, this:

<img width="933" alt="image" src="https://user-images.githubusercontent.com/25578179/42137089-3c50fe24-7d1b-11e8-99eb-c42330ef76c6.png">

It seems like this has happened to others before for no particular reason. I'm still looking into it, I'll push the changes once I find one.

austintackaberry commented:
I think you need to stage your changes to .babelrc.js, package.json, and package-lock.json

radotreyes commented:
Doesn't seem to be it:

<img width="904" alt="image" src="https://user-images.githubusercontent.com/25578179/42137324-6e8a53d2-7d1f-11e8-84d3-9a7a3b308954.png">

austintackaberry commented:
Hmmm...I made my own test branch where I added spread operator and added that line to the .babelrc and committed it, and I didn't have a problem. If you're all out of ideas, maybe just bypass the git hook, so that others can take a look and help debug?

radotreyes commented:
Just pushed the changes while bypassing the linter problem. Still trying to figure out what's causing this.

nikrb commented:
just pulled this branch to look at the issues, but it seems to work for me. Added a test for spread operator and that too seems fine.
Are you still experiencing issues with this pr?

radotreyes commented:
Just pushed the changes. Cleaned up the `Sidebar.js` tests a little to remove the dummy variable test suite, since I added the helper function that depends on the real data now. Played around to make sure nothing is wrong, and I haven't caught anything on my end. Let me know if we're good!

austintackaberry commented:
Looks good! Only thing is that the book sidebar isn't updating for me

radotreyes commented:
Noted, and found a few more bugs as well. Will have a resolution soon.
